File: 251004/Kalitsko/4-5lab/discussion/src/main/java/com/example/rest/service/Impl/PostService.java
package com.example.rest.service.Impl;

import com.example.rest.dto.InTopicDTO;
import com.example.rest.dto.OutTopicDTO;
import com.example.rest.dto.PostRequestTo;
import com.example.rest.dto.PostResponseTo;
import com.example.rest.entity.Post;
import com.example.rest.exceptionHandler.CreatorNotFoundException;
import com.example.rest.mapper.PostMapper;
import com.example.rest.repository.PostRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.support.KafkaHeaders;
import org.springframework.messaging.Message;
import org.springframework.messaging.handler.annotation.Header;
import org.springframework.messaging.handler.annotation.Payload;
import org.springframework.messaging.handler.annotation.SendTo;
import org.springframework.messaging.support.MessageBuilder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class PostService {
    private final PostRepository postRepository;
    private final PostMapper postMapper;

    @Autowired
    public PostService(PostRepository postRepository, PostMapper postMapper) {
        this.postRepository = postRepository;
        this.postMapper = postMapper;
    }

    @KafkaListener(topics = "InTopic", groupId = "posts-group")
    @SendTo
    public Message<OutTopicDTO> handlePostRequest(@Payload InTopicDTO request,
                                                     @Header(name = KafkaHeaders.REPLY_TOPIC, required = false) byte[] replyTopic,
                                                     @Header(name = KafkaHeaders.CORRELATION_ID, required = false) byte[] correlationId) {
        PostRequestTo postRequestTo = request.getPostRequestTo();
        String method = request.getMethod();

        OutTopicDTO response;

        try {
            if (method.equals("POST")) {
                handleSave(postRequestTo);
                return null;
            } else if (method.equals("GET")) {
                response = postRequestTo != null ? handleFindById(postRequestTo.getId()) : handleFindAll();
            } else if (method.equals("PUT")) {
                response = handleUpdate(postRequestTo);
            } else if (method.equals("DELETE")) {
                response = handleDelete(postRequestTo.getId());
            } else {
                response = new OutTopicDTO("Unsupported method: " + method, "DECLINE");
            }
        } catch (Exception ex) {
            response = new OutTopicDTO("Error: " + ex.getMessage(), "DECLINE");
        }

        if (replyTopic != null && correlationId != null) {
            return MessageBuilder.withPayload(response)
                    .setHeader(KafkaHeaders.TOPIC, new String(replyTopic))
                    .setHeader(KafkaHeaders.CORRELATION_ID, correlationId)
                    .build();
        } else {
            return null;
        }
    }

    private OutTopicDTO handleSave(PostRequestTo dto) {
        Post post = postMapper.toEntity(dto);
        String country = "Default";
        post.setCountry(country);
        Post savedPost = postRepository.save(post);
        return new OutTopicDTO(postMapper.toResponse(savedPost), "APPROVE");
    }

    public List<PostResponseTo> findAll() {
        return postMapper.toPostResponseList(postRepository.findAll());
    }

    private OutTopicDTO handleFindAll() {
        List<PostResponseTo> postResponseDTOList = findAll();
        return new OutTopicDTO(postResponseDTOList, "APPROVE");
    }

    public PostResponseTo findById(Long id) {
        List<Post> allPosts = postRepository.findAll();
        return allPosts.stream()
                .filter(post -> post.getId().equals(id))
                .findFirst()
                .map(postMapper::toResponse)
                .orElseThrow(() -> new CreatorNotFoundException(id));
    }

    private OutTopicDTO handleFindById(Long id) {
        try {
            return new OutTopicDTO(findById(id), "APPROVE");
        } catch (RuntimeException ex) {
            return new OutTopicDTO(ex.getMessage(), "DECLINE");
        }
    }

    private OutTopicDTO handleUpdate(PostRequestTo dto) {
        Post currPost = postRepository.findAll().stream()
                .filter(post -> post.getId().equals(dto.getId()))
                .findFirst()
                .orElseThrow(() -> new CreatorNotFoundException(dto.getId()));

        currPost.setContent(dto.getContent());
        Post updatedPost = postRepository.save(currPost);
        return new OutTopicDTO(postMapper.toResponse(updatedPost), "APPROVE");
    }

    public PostResponseTo update(PostRequestTo dto) {
        Post currPost = postRepository.findAll().stream()
                .filter(post -> post.getId().equals(dto.getId()))
                .findFirst()
                .orElseThrow(() -> new CreatorNotFoundException(dto.getId()));

        currPost.setContent(dto.getContent());
        Post updatedPost = postRepository.save(currPost);
        return postMapper.toResponse(updatedPost);
    }

    private OutTopicDTO handleDelete(Long id) {
        Post currPost = postRepository.findAll().stream()
                .filter(post -> post.getId().equals(id))
                .findFirst()
                .orElseThrow(() -> new CreatorNotFoundException(id));
        if (currPost != null) {
            postRepository.delete(currPost);
        } else {
            throw new CreatorNotFoundException(id);
        }
        return new OutTopicDTO(postMapper.toResponse(currPost), "APPROVE");
    }

}

File: 251003/Pankratiev/rest/discussion/src/main/kotlin/bsuir/dc/discussion/service/PostService.kt
package bsuir.dc.discussion.service

import bsuir.dc.discussion.dto.from.PostRequestTo
import bsuir.dc.discussion.dto.to.PostResponseTo
import bsuir.dc.discussion.entity.PostKey
import bsuir.dc.discussion.mapper.toEntity
import bsuir.dc.discussion.mapper.toResponse
import bsuir.dc.discussion.repository.PostRepository
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.stereotype.Service
import java.util.NoSuchElementException

@Service
class PostService(private val postRepository: PostRepository) {

    fun createPost(postRequestTo: PostRequestTo, country: String): PostResponseTo {
        return postRepository.save(postRequestTo.toEntity(country)).toResponse()
    }

    fun getAllPosts(): List<PostResponseTo> {
        return postRepository.findAll().map { it.toResponse() }
    }

    fun getPostById(country: String, id: Long): PostResponseTo {
        val postKey = PostKey(country, id)
        return postRepository.findById(postKey).orElseThrow {
            NoSuchElementException()
        }.toResponse()
    }

    fun deletePost(country: String, id: Long) {
        val postKey = PostKey(country, id)
        postRepository.findById(postKey).orElseThrow {
            NoSuchElementException()
        }
        postRepository.deleteById(postKey)
    }

    fun updatePost(country: String, postRequestTo: PostRequestTo): PostResponseTo {
        val postKey = PostKey(country, postRequestTo.id)
        val existingPost = postRepository.findById(postKey).orElseThrow {
            NoSuchElementException()
        }
        val updatedPost = existingPost.copy(
            issueId = postRequestTo.issueId,
            content = postRequestTo.content
        )
        return postRepository.save(updatedPost).toResponse()
    }

    @KafkaListener(topics = ["post-create"], groupId = "discussion-group")
    fun listenToPostCreate(postRequestTo: PostRequestTo) {
        createPost(postRequestTo, "Belarus")
    }

    @KafkaListener(topics = ["post-update"], groupId = "discussion-group")
    fun listenToPostUpdate(postRequestTo: PostRequestTo) {
        updatePost("Belarus", postRequestTo)
    }

    @KafkaListener(topics = ["post-delete"], groupId = "discussion-group")
    fun listenToPostDelete(id: Long) {
        deletePost("Belarus", id)
    }
}